odoo/models.py

```diff
@@ -3041,12 +3041,16 @@ def check_access_rule(self, operation):
             if len(uids) != 1 or uids[0] != self._uid:
                 raise AccessError(_('For this kind of document, you may only access records you created yourself.\n\n(Document type: %s)') % (self._description,))
         else:
-            where_clause, where_params, tables = self.env['ir.rule'].domain_get(self._name, operation)
-            if where_clause:
-                query = "SELECT %s.id FROM %s WHERE %s.id IN %%s AND " % (self._table, ",".join(tables), self._table)
-                query = query + " AND ".join(where_clause)
+            query = self.env['ir.rule'].domain_get(self._name, operation)
+            if query.where_clause:
+                table = '"%s"' % self._table
                 for sub_ids in self._cr.split_for_in_conditions(self.ids):
-                    self._cr.execute(query, [sub_ids] + where_params)
+                    chunk_query = query + Query([table], [table + '.id IN %s'], [sub_ids])
+                    from_clause, where_clause, params = chunk_query.get_sql()
+                    self.env.cr.execute(
+                        "SELECT DISTINCT({table}.id) FROM {from_clause} WHERE {where_clause}".format(
```

Comment:
Avoid `DISTINCT` in queries (potential performance issue);
it is cheaper to make a `set` with results in Python.

Reply:
I noticed that deduplication is carried out in the next call so I simply removed the DISTINCT.